### PR TITLE
Web server simplifications and handers

### PR DESCRIPTION
**Pull request by dirkx**
Proposal to simplify the authentication code in de webserver (and potentially also derisk it); and to also allow for richer authentication; such as with multiple usernames and passwords, the use of Time based one-time passwords or typical IoT bearer tokens. 

Also adds a number of extra examples - including one that uses an authentication cookie.

Has been tested on ESP-8266 and ESP-32.

### CODE REVIEW

**Comment by Jason2866** [comment]
With the changes a simple Webserver now includes mbedtls software libs. This increases code size a lot. Actually the md5 routine from mbedtls is used which is in all MCUs a routine which is embedded in ROM and does not increase code size.

**Comment by dirkx** [comment]
Ok - so your estimate is that by replacing md5 by the rom routines - we can totally remove the need for mbedtls ? Correct ? If so - happy to give that a try.

**Comment by Jason2866** [comment]
Yes, see this change https://github.com/espressif/arduino-esp32/pull/8667

File: libraries/ESP32/examples/MD5Builder/MD5Builder.ino [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+    if (memcmp(raw, res, 16)) 
+      Serial.println("Odd - failing MD5 on byte array when compared as bytes"); 
+    else
+      Serial.println("OK!");
+
+  }
+}
+
+void loop() {
+}
```
Comment (lucasssvaz) [inline libraries/ESP32/examples/MD5Builder/MD5Builder.ino:97]:
```suggestion
}

```

File: libraries/WebServer/examples/HttpAuthCallback/HttpAuthCallback.ino [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+  Serial.print("Open http://");
+  Serial.print(WiFi.localIP());
+  Serial.println("/ in your browser to see it working");
+}
+
+void loop() {
+  ArduinoOTA.handle();
+  server.handleClient();
+  delay(2);//allow the cpu to switch to other tasks
+}
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpAuthCallback/HttpAuthCallback.ino:63]:
```suggestion
}

```

File: libraries/WebServer/examples/HttpAuthCallbackInline/HttpAuthCallbackInline.ino [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+  Serial.print("Open http://");
+  Serial.print(WiFi.localIP());
+  Serial.println("/ in your browser to see it working");
+}
+
+void loop() {
+  ArduinoOTA.handle();
+  server.handleClient();
+  delay(2);//allow the cpu to switch to other tasks
+}
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpAuthCallbackInline/HttpAuthCallbackInline.ino:66]:
```suggestion
}

```

File: libraries/WebServer/examples/HttpAuthOneTimePassword/HttpAuthOneTimePassword.ino
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpAuthOneTimePassword/HttpAuthOneTimePassword.ino:1]: I think this example would be more useful/proper if it was provided by the OTP library and not the arduino core.

File: libraries/WebServer/examples/HttpAuthOneTimePasswordNaive/HttpAuthOneTimePasswordNaive.ino
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpAuthOneTimePasswordNaive/HttpAuthOneTimePasswordNaive.ino:1]: I think this example would be more useful/proper if it was provided by the OTP library and not the arduino core.

File: libraries/WebServer/examples/HttpBasicAuthSHA1/HttpBasicAuthSHA1.ino [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+  Serial.print("Open http://");
+  Serial.print(WiFi.localIP());
+  Serial.println("/ in your browser to see it working");
+}
+
+void loop() {
+  ArduinoOTA.handle();
+  server.handleClient();
+  delay(2);//allow the cpu to switch to other tasks
+}
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpBasicAuthSHA1/HttpBasicAuthSHA1.ino:75]:
```suggestion
}

```

File: libraries/WebServer/examples/HttpBasicAuthSHA1orBearerToken/HttpBasicAuthSHA1orBearerToken.ino [hunk 5]
```diff
@@ -0,0 +1,109 @@
+#include <WiFi.h>
+#include <ESPmDNS.h>
+#include <ArduinoOTA.h>
+#include <WebServer.h>
+#include "mbedtls/sha1.h"
+
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpBasicAuthSHA1orBearerToken/HttpBasicAuthSHA1orBearerToken.ino:6]:
```suggestion
```

File: libraries/WebServer/examples/HttpBasicAuthSHA1orBearerToken/HttpBasicAuthSHA1orBearerToken.ino [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+
+/// We have two options - we either come in with a bearer
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpBasicAuthSHA1orBearerToken/HttpBasicAuthSHA1orBearerToken.ino:8]:
```suggestion
//  We have two options - we either come in with a bearer
```

File: libraries/WebServer/examples/HttpBasicAuthSHA1orBearerToken/HttpBasicAuthSHA1orBearerToken.ino [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  Serial.print("Open http://");
+  Serial.print(WiFi.localIP());
+  Serial.println("/ in your browser to see it working");
+}
+
+void loop() {
+  ArduinoOTA.handle();
+  server.handleClient();
+  delay(2);//allow the cpu to switch to other tasks
+}
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpBasicAuthSHA1orBearerToken/HttpBasicAuthSHA1orBearerToken.ino:109]:
```suggestion
}

```

File: libraries/WebServer/examples/HttpBasicAuthSHA1orBearerToken/HttpBasicAuthSHA1orBearerToken.ino [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+#include "mbedtls/sha1.h"
+
+
+/// We have two options - we either come in with a bearer
+// token - i.e. a special header or API token; or we
+// get a normal HTTP style basic auth prompt.
+//
+// To do a bearer fetch - use something like Swagger or with curl:
+//
+//    curl https://myesp.com/  -H "Authorization: Bearer SecritToken"
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpBasicAuthSHA1orBearerToken/HttpBasicAuthSHA1orBearerToken.ino:14]: Using this command with the proper IP makes the application crash with a Load Exception on `WebServer::authenticateBasicSHA1`

File: libraries/WebServer/src/WebServer.cpp [hunk 9]
```diff
@@ -23,14 +23,17 @@
...
 #include "WiFiServer.h"
 #include "WiFiClient.h"
 #include "WebServer.h"
+#include "HEXBuilder.h"
 #include "FS.h"
 #include "detail/RequestHandlersImpl.h"
 #include "mbedtls/md5.h"
-
+#include "mbedtls/sha1.h"
+#include "mbedtls/base64.h"
```
Comment (lucasssvaz) [inline libraries/WebServer/src/WebServer.cpp:36]: As @Jason2866 said, I don't think it is a good idea to the WebServer to depend on mbedtls. Maybe a minimal version of the functions needed could be provided ?

File: cores/esp32/HEXBuilder.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+  Lesser General Public License for more details.
+
+  You should have received a copy of the GNU Lesser General Public
+  License along with this library; if not, write to the Free Software
+  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+*/
+#ifndef __ESP8266_HEX_BUILDER__
+#define __ESP8266_HEX_BUILDER__
```
Comment (lucasssvaz) [inline cores/esp32/HEXBuilder.h:20]:
```suggestion
#ifndef HEXBuilder_h
#define HEXBuilder_h
```

File: libraries/WebServer/examples/HttpAuthCallback/HttpAuthCallback.ino [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+{
+  for (int i = 0; i < N_CREDENTIALS; i++) {
+    if (username == passwdfile[i].username)
+      return new String(passwdfile[i].password);
+  }
+  return NULL;
+}
+
+void setup() {
+  Serial.begin(115200);
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpAuthCallback/HttpAuthCallback.ino:35]:
```suggestion
  Serial.begin(115200);
  while (!Serial) { delay(10); }
```

File: libraries/WebServer/examples/HttpAuthCallbackInline/HttpAuthCallbackInline.ino [hunk 12]
```diff
@@ -0,0 +1,66 @@
...
+  {  "admin",  "esp32" },
+  {  "fred",  "41234123" },
+  {  "charlie",  "sdfsd" },
+  {  "alice",  "vambdnkuhj" },
+  {  "bob",  "svcdbjhws12" },
+  { NULL, NULL }
+};
+
+void setup() {
+  Serial.begin(115200);
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpAuthCallbackInline/HttpAuthCallbackInline.ino:26]:
```suggestion
  Serial.begin(115200);
  while (!Serial) { delay(10); }
```

File: libraries/WebServer/examples/HttpBasicAuthSHA1/HttpBasicAuthSHA1.ino [hunk 13]
```diff
@@ -0,0 +1,75 @@
...
+//
+const char* www_username_hex = "hexadmin";
+const char* www_password_hex = "8cb124f8c277c16ec0b2ee00569fd151a08e342b";
+
+// The same; but now expressed as a base64 string (e.g. as commonly used
+// by webservers). Created by ` echo -n esp32 | openssl sha1 -binary | openssl base64`
+//
+const char* www_username_base64 = "base64admin";
+const char* www_password_base64 = "jLEk+MJ3wW7Asu4AVp/RUaCONCs=";
+
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpBasicAuthSHA1/HttpBasicAuthSHA1.ino:34]:
```suggestion
```

File: libraries/WebServer/examples/HttpBasicAuthSHA1/HttpBasicAuthSHA1.ino [hunk 14]
```diff
@@ -0,0 +1,75 @@
...
+
+void setup() {
+  Serial.begin(115200);
```
Comment (lucasssvaz) [inline libraries/WebServer/examples/HttpBasicAuthSHA1/HttpBasicAuthSHA1.ino:37]:
```suggestion
  Serial.begin(115200);
  while (!Serial) { delay(10); }
```

File: libraries/ESP32/examples/HEXBuilder/HEXBuilder.ino [hunk 15]
```diff
@@ -0,0 +1,76 @@
+#include <HEXBuilder.h>
+
+void setup() {
+  Serial.begin(115200);
+  delay(100);
```
Comment (lucasssvaz) [inline libraries/ESP32/examples/HEXBuilder/HEXBuilder.ino:5]:
```suggestion
  while (!Serial) { delay(10); }
```

File: libraries/ESP32/examples/MD5Builder/MD5Builder.ino [hunk 16]
```diff
@@ -0,0 +1,97 @@
...
+// source code.
+//
+// MD5Builder helps you obfuscate this (it is not terribly secure, MD5
+// has been phased out as insecure eons ago) by letting you create an
+// MD5 of the data the user entered; and then compare this to an MD5
+// string that you have put in your code.
+//
+void setup() {
+  Serial.begin(115200);
+  delay(100);
```
Comment (lucasssvaz) [inline libraries/ESP32/examples/MD5Builder/MD5Builder.ino:15]:
```suggestion
  while (!Serial) { delay(10); }
```

File: libraries/ESP32/examples/MD5Builder/MD5Builder.ino [hunk 17]
```diff
@@ -0,0 +1,97 @@
...
+// has entered to a build in string. However this means that the
+// password ends up `in the clear' in the firmware and in your
+// source code.
+//
+// MD5Builder helps you obfuscate this (it is not terribly secure, MD5
+// has been phased out as insecure eons ago) by letting you create an
+// MD5 of the data the user entered; and then compare this to an MD5
+// string that you have put in your code.
+//
+void setup() {
```
Comment (lucasssvaz) [inline libraries/ESP32/examples/MD5Builder/MD5Builder.ino:13]:
```suggestion

void setup() {
```

**Review by lucasssvaz — CHANGES_REQUESTED** [review]
I think this is a very nice improvement but needs some small fixes first, mainly the mbedtls overhead and the crashing when using a bearer. Please also fix the conflicts to update do 3.0.0

**Review by lucasssvaz — COMMENTED** [review]
@SuGlider @P-R-O-C-H-Y PR fixed and ready for review

**Review by lucasssvaz — APPROVED** [review]

File: libraries/WebServer/src/WebServer.cpp [hunk 18]
```diff
@@ -128,91 +129,185 @@ static String md5str(String &in){
...
-      log_v("Hash of GET:uri=%s", _H2.c_str());
-      String _responsecheck = "";
-      if(authReq.indexOf(FPSTR(qop_auth)) != -1 || authReq.indexOf(FPSTR(qop_auth_quoted)) != -1) {
-          _responsecheck = md5str(_H1 + ':' + _nonce + ':' + _nc + ':' + _cnonce + F(":auth:") + _H2);
-      } else {
-          _responsecheck = md5str(_H1 + ':' + _nonce + ':' + _H2);
-      }
-      log_v("The Proper response=%s", _responsecheck.c_str());
-      if(_response == _responsecheck){
+bool WebServer::authenticateBasicSHA1(const char * _username, const char * _sha1Base64orHex) {
```
Comment (me-no-dev) [inline libraries/WebServer/src/WebServer.cpp:132]: this method seems to have indentation issues. There are some more throughout the file as well, but fix only what has been changed

File: libraries/WebServer/src/WebServer.cpp [hunk 19]
```diff
@@ -128,91 +129,185 @@ static String md5str(String &in){
...
+
+    return ((username.equalsConstantTime(_username)) &&
+            (String((char*)sha1calc).equalsConstantTime(_sha1Base64orHex)) &&
+            (mode == BASIC_AUTH) /* to keep things somewhat time constant. */
+            ) ? new String(params[0]) : NULL;
+});
+}
+
+bool WebServer::authenticate(const char * _username, const char * _password){
+	return WebServer::authenticate([_username,_password](HTTPAuthMethod mode, String username, String params[])->String * {
```
Comment (me-no-dev) [inline libraries/WebServer/src/WebServer.cpp:180]: indentation

**Pull request merged**